node/test_helpers.go

```diff
@@ -7,10 +7,21 @@ import (
 	tmtypes "github.com/tendermint/tendermint/types"
 
 	"github.com/rollkit/rollkit/conv"
+	"github.com/rollkit/rollkit/mocks"
 )
 
 var genesisValidatorKey = ed25519.GenPrivKey()
 
+func countFunctionCalls(app *mocks.Application, functionName string) uint {
```

Comment:
Even though this *should* be thread safe as we are calling it synchronously, I think we can do better. 

We should try and avoid directly accessing fields of structs that are controlled by mutexes in external packages. 

The crux here is that the mocks package uses the assert package and will fail if a condition isn't met, and there isn't a concept of just checking a value. So we need to handle that on our side. 

We can do this will a simple wrapper that catches the `Fail` and `Errorf` calls from the assert package.

```go
// This could be called something more explicit like, ExpectFailTester or RetryTester so that it is more clear when we should use it. 
type MockTester struct {
	t *testing.T
}

func (m *MockTester) Fail() {
	fmt.Println("Failed")
}

func (m *MockTester) Errorf(format string, args ...interface{}) {
	fmt.Printf(format, args...)
}
```

If you use this in a example like below, the output will log the invalid result, but the test itself registers a pass
```go
func TestMockTester(t *testing.T) {
	m := &MockTester{t: t}
	assert := assert.New(m)
        assert.False(true)
}

// Output
go test -v -run=TestMockTester
=== RUN   TestMockTester

        Error Trace:   .....
        Error:          Should be false
--- PASS: TestMockTester (0.00s)
PASS
ok      ./././ 0.444s
```


So how does this help use. This means we can remove this function update [here](https://github.com/rollkit/rollkit/pull/992/files#diff-7af4c73284d8c84344fe9b93472357f0318453ef83c0c30370343856b2d5a985R573) and do something like this which would be threadsafe:
```go
go func() {
		defer close(doneChan)
                // Use Mock Tester to bypass test failures for retry
	        m := &MockTester{t: t}
		require.NoError(t, testutils.Retry(100, 100*time.Millisecond, func() error {
			for i := 0; i < len(apps); i++ {
                                 // Since we are by passing the test failure, we can check the return value, which is a bool
                                 if !apps[i].AssertCalled(m, "DeliverTx") {
					return errors.New("DeliverTx hasn't been called yet")
				}
			}
			return nil
		}))
	}()

Reply:
implemented